thrust/thrust/iterator/shuffle_iterator.h

```diff
@@ -0,0 +1,184 @@
+                     IndexType,
+                     IndexType,
+                     system,
+                     traversal,
+                     IndexType,
+                     difference>;
+};
+} // namespace detail
+
+//! \p shuffle_iterator is an iterator which generates a sequence of values representing a random permutation.
+//! \tparam IndexType The type of the index to shuffle.
+//! \tparam BijectionFunc The bijection to use. This should be a bijective function that maps [0..n) -> [0..n). It must
+//! be deterministic and stateless. The function will be constructed with parameters (n, g) where n is the number of
+//! elements in the permutation and g is a \c URBG that can be used to seed the bijection.
+//!
+//! \addtogroup iterators
+//! \{
+
+//! \addtogroup fancyiterator Fancy Iterators
+//! \ingroup iterators
+//! \{
+
+//! \p shuffle_iterator is an iterator which generates a sequence of values representing a random permutation. This
+//! iterator is useful for working with random permutations of a range without explicitly storing it in memory. The
+//! shuffle iterator is also useful for sampling from a range by selecting only a subset of the elements in the
+//! permutation.
+//!
+//! The following code snippet demonstrates how to create a \p shuffle_iterator which generates a random permutation of
+//! a vector.
+//!
+//! \code
+//! #include <thrust/iterator/shuffle_iterator.h>
+//! ...
+//! // create a shuffle iterator
+//! thrust::shuffle_iterator<int> iterator(4, thrust::default_random_engine(0xDEADBEEF));
+//! iterator[0] // returns 1
+//! iterator[1] // returns 3
+//! iterator[2] // returns 2
+//! iterator[3] // returns 0
+//!
+//! thrust::device_vector<int> vec = {0, 10, 20, 30};
+//! thrust::device_vector<int> shuffled(4);
+//! thrust::gather(iterator, iterator + 4, vec.begin(), shuffled.begin());
+//! shuffled // returns {10, 30, 20, 0}
+//! \endcode
+//!
+//! This next example demonstrates how to use a \p shuffle_iterator to randomly sample from a vector.
+//!
+//! \code
+//! #include <thrust/iterator/shuffle_iterator.h>
+//! ...
+//! // create a shuffle iterator
+//! thrust::shuffle_iterator<int> iterator(100, thrust::default_random_engine(0xDEADBEEF));
+//!
+//! iterator[0] // returns 38
+//! iterator[1] // returns 50
+//! iterator[2] // returns 18
+//! iterator[3] // returns 12
+//!
+//! // create a vector of size 100
+//! thrust::device_vector<int> vec(100);
+//! thrust::device_vector<int> sample(4);
+//!
+//! // fill vec with random values
+//! thrust::sequence(vec.begin(), vec.end(), 100);
+//!
+//! // sample 4 random values from vec
+//! thrust::gather(iterator, iterator + 4, vec.begin(), sample.begin());
+//! sample // returns {138, 150, 118, 112}
+//! \endcode
+//!
+//! \see make_shuffle_iterator
+template <class IndexType, class BijectionFunc = thrust::detail::random_bijection<IndexType>>
+class shuffle_iterator : public detail::make_shuffle_iterator_base<IndexType, BijectionFunc>::type
```

Comment:
In the issue I mentioned this could be implemented as a composition of a counting iterator and transform iterator.
I have implemented this as a standalone iterator as that seems more in keeping with the design of the rest of the thrust iterators. However, implementing as a composition would lower the maintenance burden if it means we automatically inherit any features of transform/counting iterator when they are added.

Reply:
I think we're unlikely to add new features to those iterators, keeping this standalone is fine with me.